File: hartshorn-integration-tests/src/integration-test/java/test/org/dockbox/hartshorn/inject/compatibility/JavaxCompatibilityTests.java
/*
 * Copyright 2019-2025 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package test.org.dockbox.hartshorn.inject.compatibility;

import org.dockbox.hartshorn.inject.ComponentResolutionException;
import org.dockbox.hartshorn.inject.processing.construction.AnnotatedMethodComponentPostConstructor;
import org.dockbox.hartshorn.inject.provider.HierarchicalComponentProviderOrchestrator;
import org.dockbox.hartshorn.inject.provider.MissingInjectConstructorException;
import org.dockbox.hartshorn.inject.targets.MethodsAndFieldsInjectionPointResolver;
import org.dockbox.hartshorn.launchpad.ApplicationContext;
import org.dockbox.hartshorn.launchpad.HartshornApplicationConfigurer;
import org.dockbox.hartshorn.launchpad.SimpleApplicationContext;
import org.dockbox.hartshorn.launchpad.environment.ConfigurableApplicationEnvironment;
import org.dockbox.hartshorn.test.TestApplicationCustomizer;
import org.dockbox.hartshorn.test.junit.HartshornIntegrationTest;
import org.dockbox.hartshorn.util.configure.ContextualInitializer;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import javax.annotation.Resource;
import javax.inject.Inject;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

@HartshornIntegrationTest(
    includeBasePackages = false,
    customizers = DisableSingleConstructorFallbackTestCustomizer.class
)
class JavaxCompatibilityTests {

    /**
     * Test application customizer that enables support for Javax annotations in the test
     * application. Note that typically Javax support would be enabled through
     * {@link HartshornApplicationConfigurer#withJavaxAnnotations()}.
     */
    public static class EnableJavaxTestApplicationCustomizer implements TestApplicationCustomizer {
        @Override
        public void customizeApplication(SimpleApplicationContext.Configurer configurer) {
            configurer.componentProvider(HierarchicalComponentProviderOrchestrator.create(
                orchestrator -> orchestrator.componentPostConstructor(
                        AnnotatedMethodComponentPostConstructor.create(
                                AnnotatedMethodComponentPostConstructor.Configurer
                                        ::withJavaxAnnotations
                        )
                )));
        }

        @Override
        public void customizeEnvironment(ConfigurableApplicationEnvironment.Configurer configurer) {
            configurer.injectionPointsResolver(ContextualInitializer.defer(() ->
                MethodsAndFieldsInjectionPointResolver.create(
                    MethodsAndFieldsInjectionPointResolver.Configurer::withJavaxAnnotations
                )
            ));
        }
    }

    // Hartshorn annotation, as Jakarta is disabled for several tests
    @org.dockbox.hartshorn.inject.annotations.Inject
    private ApplicationContext applicationContext;

    @Test
    @HartshornIntegrationTest(customizers = EnableJavaxTestApplicationCustomizer.class)
    @DisplayName("Javax annotations are supported for field injection, if enabled")
    void javaxFieldInjectSupportedIfEnabled() {
        this.applicationContext.defaultBinder().bind(String.class).singleton("Hello, World!");

        FieldJavaxComponent component =
            this.applicationContext.defaultProvider().get(FieldJavaxComponent.class);
        assertThat(component).isNotNull();

        assertThat(component.messageAsInject).isEqualTo("Hello, World!");
        assertThat(component.messageAsResource).isEqualTo("Hello, World!");
    }

    @Test
    @DisplayName("Javax annotations are not supported for field injection, if disabled")
    void javaxFieldInjectFailsIfDisabled() {
        this.applicationContext.defaultBinder().bind(String.class).singleton("Hello, World!");

        FieldJavaxComponent component =
            this.applicationContext.defaultProvider().get(FieldJavaxComponent.class);
        assertThat(component).isNotNull();

        assertThat(component.messageAsInject).isNull();
        assertThat(component.messageAsResource).isNull();
    }

    @Test
    @HartshornIntegrationTest(customizers = EnableJavaxTestApplicationCustomizer.class)
    @DisplayName("Javax annotations are supported for constructor injection, if enabled")
    void javaxConstructorInjectSupportedIfEnabled() {
        this.applicationContext.defaultBinder().bind(String.class).singleton("Hello, World!");

        ConstructorJavaxComponent component =
            this.applicationContext.defaultProvider().get(ConstructorJavaxComponent.class);
        assertThat(component).isNotNull();

        assertThat(component.messageAsInject).isEqualTo("Hello, World!");
    }

    @Test
    @DisplayName("Javax annotations are not supported for constructor injection, if disabled")
    void javaxConstructorInjectFailsIfDisabled() {
        this.applicationContext.defaultBinder().bind(String.class).singleton("Hello, World!");

        // No compatible constructor should be found, as Javax compatibility is disabled
        Throwable componentResolutionException = assertThatThrownBy(() -> {
            this.applicationContext.defaultProvider()
                .get(JavaxCompatibilityTests.ConstructorJavaxComponent.class);
        }).isInstanceOf(ComponentResolutionException.class).actual();
        assertThat(componentResolutionException.getCause())
                .isInstanceOf(MissingInjectConstructorException.class);
    }

    @Test
    @HartshornIntegrationTest(customizers = EnableJavaxTestApplicationCustomizer.class)
    @DisplayName("Javax annotations are supported for method injection, if enabled")
    void javaxMethodInjectSupportedIfEnabled() {
        this.applicationContext.defaultBinder().bind(String.class).singleton("Hello, World!");

        MethodJavaxComponent component =
            this.applicationContext.defaultProvider().get(MethodJavaxComponent.class);
        assertThat(component).isNotNull();

        assertThat(component.messageAsInject).isEqualTo("Hello, World!");
        assertThat(component.messageAsResource).isEqualTo("Hello, World!");
    }

    @Test
    @DisplayName("Javax annotations are not supported for method injection, if disabled")
    void javaxMethodInjectFailsIfDisabled() {
        this.applicationContext.defaultBinder().bind(String.class).singleton("Hello, World!");

        MethodJavaxComponent component =
            this.applicationContext.defaultProvider().get(MethodJavaxComponent.class);
        assertThat(component).isNotNull();

        assertThat(component.messageAsInject).isNull();
        assertThat(component.messageAsResource).isNull();
    }

    private static class FieldJavaxComponent {
        @Inject
        String messageAsInject;

        @Resource
        String messageAsResource;
    }

    private static class ConstructorJavaxComponent {
        final String messageAsInject;

        @Inject
        public ConstructorJavaxComponent(String messageAsInject) {
            this.messageAsInject = messageAsInject;
        }
    }

    private static class MethodJavaxComponent {
        String messageAsInject;
        String messageAsResource;

        @Inject
        public void setMessageAsInject(String messageAsInject) {
            this.messageAsInject = messageAsInject;
        }

        @Resource
        public void setMessageAsResource(String messageAsResource) {
            this.messageAsResource = messageAsResource;
        }
    }
}

File: hartshorn-launchpad/src/test/kotlin/test/org/dockbox/hartshorn/core/compat/JavaxCompatibilityTests.kt
/*
 * Copyright 2019-2024 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package test.org.dockbox.hartshorn.core.compat

import org.dockbox.hartshorn.launchpad.ApplicationContext
import org.dockbox.hartshorn.application.createApplication
import org.dockbox.hartshorn.inject.graph.support.ComponentInitializationException
import org.dockbox.hartshorn.inject.provider.MissingInjectConstructorException
import org.dockbox.hartshorn.inject.binding.bind
import org.dockbox.hartshorn.util.TypeUtils
import org.junit.Test
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertInstanceOf
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.fail
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.assertThrows

class JavaxCompatibilityTests {

  private fun createApplicationContext(withJavax: Boolean): ApplicationContext =
    createApplication<JavaxCompatibilityTests>().initialize {
      if (withJavax) withJavaxAnnotations()
      includeBasePackages(false)
    }

  @Test
  @DisplayName("Javax annotations are supported for field injection, if enabled")
  fun testJavaxFieldInjectSupportedIfEnabled() {
    val context = this.createApplicationContext(withJavax = true)
    context.bind<String>().singleton("Hello, World!")

    val service = context.get(FieldJavaxService::class.java)
    assertNotNull(service)

    assertEquals("Hello, World!", service.messageAsInject)
    assertEquals("Hello, World!", service.messageAsResource)
  }

  @Test
  @DisplayName("Javax annotations are not supported for field injection, if disabled")
  fun testJavaxFieldInjectFailsIfDisabled() {
    val context = this.createApplicationContext(withJavax = false)
    context.bind<String>().singleton("Hello, World!")

    val service = context.get(FieldJavaxService::class.java)
    assertNotNull(service)

    assertNull(service.messageAsInject)
    assertNull(service.messageAsResource)
  }

  @Test
  @DisplayName("Javax annotations are supported for constructor injection, if enabled")
  fun testJavaxConstructorInjectSupportedIfEnabled() {
    val context = this.createApplicationContext(withJavax = true)
    context.bind<String>().singleton("Hello, World!")

    val service = context.get(InjectConstructorJavaxService::class.java)
    assertNotNull(service)

    assertEquals("Hello, World!", service.messageAsInject)
  }

  @Test
  @DisplayName("Javax annotations are not supported for constructor injection, if disabled")
  fun testJavaxConstructorInjectFailsIfDisabled() {
    val context = this.createApplicationContext(withJavax = false)
    context.bind<String>().singleton("Hello, World!")

    val initializationException = assertThrows<ComponentInitializationException> {
      context.get(InjectConstructorJavaxService::class.java)
    }
    val rootCause = TypeUtils.getRootCause(initializationException)
    assertInstanceOf(MissingInjectConstructorException::class.java, rootCause)
  }

  @Test
  @DisplayName("Javax annotations are supported for method injection, if enabled")
  fun testJavaxMethodInjectSupportedIfEnabled() {
    val context = this.createApplicationContext(withJavax = true)
    context.bind<String>().singleton("Hello, World!")

    val service = context.get(MethodJavaxService::class.java)
    assertNotNull(service)

    assertEquals("Hello, World!", service.messageAsInject)
    assertEquals("Hello, World!", service.messageAsResource)
  }

  @Test
  @DisplayName("Javax annotations are not supported for method injection, if disabled")
  fun testJavaxMethodInjectFailsIfDisabled() {
    val context = this.createApplicationContext(withJavax = false)
    context.bind<String>().singleton("Hello, World!")

    val service = context.get(MethodJavaxService::class.java)
    assertNotNull(service)

    assertNull(service.messageAsInject)
    assertNull(service.messageAsResource)
  }

  class FieldJavaxService {

    @javax.inject.Inject
    internal var messageAsInject: String? = null

    @javax.annotation.Resource
    internal var messageAsResource: String? = null
  }

  class InjectConstructorJavaxService {
    val messageAsInject: String

    @javax.inject.Inject
    constructor(messageAsInject: String) {
      this.messageAsInject = messageAsInject
    }

    constructor(markerToVerify: Any) {
      this.messageAsInject = ""
      fail<Void>("This constructor should never be called")
    }
  }

  class MethodJavaxService {

    internal var messageAsInject: String? = null
    internal var messageAsResource: String? = null

    @javax.inject.Inject
    fun setMessageAsInject(messageAsInject: String) {
      this.messageAsInject = messageAsInject
    }

    @javax.annotation.Resource
    fun setMessageAsResource(messageAsResource: String) {
      this.messageAsResource = messageAsResource
    }
  }
}
